### ISSUE TITLE
support tileserver subdomains with curly-brace syntax

### ISSUE THREAD

**Issue by meetar** [tangrams/tangram#445]
Would like to be able to specify a range of tileserver subdomains, eg with the syntax:

`http://osm2vectortiles-{0-3}.tileserver.com/v1/{z}/{x}/{y}.pbf`

**Comment by ThomasG77** [comment]
Thanks for opening the PR.

Be cautious. The common way to use subdomains is {s} where s goes from a to c.

The syntax I asked and you reported is the OpenLayers 3 where:

`http://osm2vectortiles-{0-3}.tileserver.com/v1/{z}/{x}/{y}.pbf` is equivalent to `['http://osm2vectortiles-0.tileserver.com/v1/{z}/{x}/{y}.pbf', 'http://osm2vectortiles-1.tileserver.com/v1/{z}/{x}/{y}.pbf', 'http://osm2vectortiles-2.tileserver.com/v1/{z}/{x}/{y}.pbf', 'http://osm2vectortiles-3.tileserver.com/v1/{z}/{x}/{y}.pbf'] `

`http://osm2vectortiles-{a-c}.tileserver.com/v1/{z}/{x}/{y}.pbf` is equivalent to `['http://osm2vectortiles-a.tileserver.com/v1/{z}/{x}/{y}.pbf', 'http://osm2vectortiles-b.tileserver.com/v1/{z}/{x}/{y}.pbf', 'http://osm2vectortiles-c.tileserver.com/v1/{z}/{x}/{y}.pbf'] `

**Comment by mattrei** [comment]
Hi,
this does not work by providing a {s} in the url. So is this feature now supported in Tangram, or am I doing something wrong?

Best

**Comment by bcamper** [comment]
Hi @mattrei - sorry, it's not officially supported yet, we need to settle on syntax. (Turns out we have a very early undocumented version of this in Tangram JS, but we need to agree on syntax for parity with our Tangram ES native code base.)

**Comment by bcamper** [comment]
@matteblair here is some old, undocumented syntax that Tangram JS implements (but which we have never used):

`http://{s:[a, b, c, d]}.some-tile-server.com/{z}/{x}/{y}.pbf`

The relevant portion here is `{s:[a, b, c, d]}`, which allows a list of subdomain strings to be specified in an array (requests to the data source then round-robin through each subdomain). This requires each subdomain to be explicitly listed, rather than automatically expanded from shorthand like `0-3` or `a-d`. Since there appears to be no single standard in the community, the idea here was to make something flexible but more explicit. An alternative would be to require the `url` source parameter to be an array listing out all the full URLs to use, which requires the user to be even more explicit, and is more error prone to copy-paste errors, but requires less special syntax handling and is less error-prone to syntax mishaps.

Any thoughts?

**Comment by matteblair** [comment]
Ah interesting!

I like the syntax style that leaflet uses for this, with `{s}` as a subdomain placeholder and a separate parameter like `url_subdomains` containing the list of replacements. That seems to involve the least amount of manual parsing. e.g.
```yaml
url: https://{s}.som-tile-server.com/{z}/{y}/{x}.pbf
url_subdomains: [a, b, c]
```

That said, how much of an impact on performance does this behavior have? It seems like this pattern is becoming less popular.

**Comment by ThomasG77** [comment]
Not using the scheme has direct impact on performance. See this old article http://ajaxian.com/archives/using-cnames-to-get-around-browser-connection-limits
The browser connection limits remain e.g http://www.browserscope.org/?category=network&v=top, at least for http 1.x. Don't know exactly the impact of subdomain when using http2

**Comment by bcamper** [comment]
@ThomasG77 my understanding is that browsers have become much smarter at throttling and prioritizing traffic, such that this technique is less effective than it was several years ago. For example, from a quick check in Chrome and FF, it doesn't appear that they use multiple hosts. I can't say definitively though without more research.

That said, I'm fine supporting it given the relatively low complexity and prevalence of the pattern.

**Comment by bcamper** [comment]
@matteblair I like that syntax and it should be easy to implement.

**Comment by matteblair** [comment]
Cool, let me know if I've overlooked anything in my issue description for tangram-es: https://github.com/tangrams/tangram-es/issues/1386

**Comment by bcamper** [comment]
Pull request for this: #525

**Comment by pnorman** [comment]
> @ThomasG77 my understanding is that browsers have become much smarter at throttling and prioritizing traffic, such that this technique is less effective than it was several years ago. For example, from a quick check in Chrome and FF, it doesn't appear that Google Maps uses multiple hosts. I can't say definitively though without more research.

It depends on if the client and server support HTTP/2. If they don't, nothing much has changed.

**Comment by bcamper** [comment]
Added in #525 for upcoming v0.12.0 release (will close this issue upon release).

**Comment by dmvaldman** [comment]
Have we noticed a network speedup as a result?

**Comment by bcamper** [comment]
We don't support multiple subdomains on Mapzen tiles (maybe we will now) so I haven't benchmarked on them. Mostly we're adding this because it's a longstanding pattern in the community and lets people take advantage of tile sources that already have it setup. I would like to benchmark to find out how much it helps with modern browsers though :)

**Comment by pnorman** [comment]
> We don't support multiple subdomains on Mapzen tiles (maybe we will now) so I haven't benchmarked on them. Mostly we're adding this because it's a longstanding pattern in the community and lets people take advantage of tile sources that already have it setup. I would like to benchmark to find out how much it helps with modern browsers though :)

I did some testing with my vector tiles. Although not fully rigorous, the pattern was similar to raster tiles. Having a single subdomain can be a significant performance limitation with parallel browser connections on HTTP 1.1. Exact differences depend on the area you're looking at, how much is in cache, network, and the balance of tile sizes, but in the right conditions I got an order of magnitude speedup in time to mostly loaded when I got rid of the parallel connection limits.

tile.mapzen.com still uses HTTP/1.1, so you could test this with Firefox and adjusting `network.http.max-persistent-connections-per-server` in `about:config`.

**Comment by bcamper** [comment]
Released in [v0.12.0](https://github.com/tangrams/tangram/releases/tag/v0.12.0)
